dtfs-central-api/src/v1/routes/middleware/payload-validation/create-validation-middleware-for-schema.ts

```diff
@@ -0,0 +1,18 @@
+import { AnyZodObject, ZodError } from 'zod';
+import { RequestHandler } from 'express';
+import { HttpStatusCode } from 'axios';
+
+const getFormattedZodErrors = (error: ZodError): string[] => error.issues.map(({ path, message, code }) => `${path.join('.')}: ${message} (${code})`);
+
+export const createValidationMiddlewareForSchema =
+  <TSchema extends AnyZodObject>(schema: TSchema): RequestHandler =>
+  (req, res, next) => {
+    const { success, error, data } = schema.safeParse(req.body);
+    if (success) {
+      req.body = data;
+      return next();
+    }
+    const formattedErrors = getFormattedZodErrors(error);
+    console.error('Payload validation error occurred:', formattedErrors);
+    return res.status(HttpStatusCode.BadRequest).send(formattedErrors);
```

Comment:
interested why youve gone for safe parse as apposed to try catch (not saying either is right or wrong) -- for the code simplicity?

Reply:
Mainly for readability. I take the view that try-catch blocks are usually quite tough to read because the flow of the code gets disrupted if something you call throws an error. For example, between the two below blocks of code, I would probably choose the former:

```typescript
const amendment = await TfmFacilityRepo.findAmendmentsByAmendmentId('abc123');
if (!amendment) {
  return res.status(404).send("Failed to find an amendment with id 'abc123'");
}
return res.status(200).send(amendment);
```

```typescript
try {
  const amendment = await TfmFacilityRepo.findAmendmentsByAmendmentId('abc123');
  if (!amendment) {
    throw new NotFoundError("Failed to find an amendment with id 'abc123'");
  }
  return res.status(200).send(amendment);
} catch (error: NotFoundError) { // we know the error type here
  const { status, message } = error;
  return res.status(status).send(message);
}
``` 

I think the utility of try-catch starts to come when you have functions which can throw many different types of errors and you want to put the handling of those errors into one place. At that point, it makes more sense to me to lose the readability of more explicit checks for the benefit of easier to read and maintain error handling. In the case where we only have one check here though, I think having code which can be easily read line-by-line is better